services/users.js

```diff
@@ -82,8 +83,246 @@ module.exports = class UsersService {
     }
   }
 
+  static async setSuspensionStatus(id, until, assignedBy = null, message) {
+    let user = await UserModel.findOneAndUpdate(
+      {id},
+      {
+        $set: {
+          'status.suspension.until': until,
+        },
+        $push: {
+          'status.suspension.history': {
+            until,
+            assigned_by: assignedBy,
+            message,
+            created_at: Date.now(),
+          },
+        },
+      },
+      {
+        new: true,
+        runValidators: true,
+      }
+    );
+    if (user === null) {
+      user = await UserModel.findOne({id});
+      if (user === null) {
+        throw errors.ErrNotFound;
+      }
+
+      // Date comparisons are difficult when using MongoDB. Javascript will
+      // store the date at a higher precision than Mongo can store, hence we
+      // check if the date is within 1 second of the time we're trying to set.
+      if (
+        user.status.suspension.until === until ||
+        (user.status.suspension.until.getTime() > until.getTime() - 1000 &&
+          user.status.suspension.until.getTime() < until.getTime() + 1000)
+      ) {
+        return user;
+      }
+
+      throw new Error(
+        'suspension status change edit failed for an unknown reason'
+      );
+    }
+
+    // Emit that the user username status was changed.
+    await events.emitAsync(USERS_SUSPENSION_CHANGE, user, {
+      until,
+      message,
+      assignedBy,
+    });
+
+    return user;
+  }
+
+  static async setBanStatus(id, status, assignedBy = null, message) {
+    let user = await UserModel.findOneAndUpdate(
+      {
+        id,
+        status: {
+          $ne: status,
```

Comment:
Is this query here correct? looks odd

Reply:
This prevents a user (who is currently banned) from being banned again. As each time we push the item into history. Without, it would look like:

```js
[
  {status: BANNED},
  {status: BANNED},
  {status: ACTIVE}
]
```

Now we ensure that the user doesn't already have that status first.

Reply:
But status is no longer a string, but an object of statuses

```js
  // Status stores the user status information regarding permissions,
  // capabilities and moderation state.
  status: {

    // Username stores the current user status for the username as well as the
    // history of changes.
    username: {

      // Status stores the current username status.
      status: {
        type: String,
        enum: USER_STATUS_USERNAME,
      },

      // History stores the history of username status changes.
      history: [{

        // Status stores the historical username status.
        status: {
          type: String,
          enum: USER_STATUS_USERNAME,
        },

        // assigned_by stores the user id of the user who assigned this status.
        assigned_by: {type: String, default: null},

        // created_at stores the date when this status was assigned.
        created_at: {type: Date, default: Date.now}
      }],
    },

    // Banned stores the current user banned status as well as the history of
    // changes.
    banned: {

      // Status stores the current user banned status.
      status: {
        type: Boolean,
        required: true,
        default: false,
      },
      history: [{

        // Status stores the historical banned status.
        status: Boolean,

        // assigned_by stores the user id of the user who assigned this status.
        assigned_by: {type: String, default: null},

        // message stores the email content sent to the user.
        message: {type: String, default: null},

        // created_at stores the date when this status was assigned.
        created_at: {type: Date, default: Date.now}
      }],
    },

    // Suspension stores the current user suspension status as well as the
    // history of changes.
    suspension: {

      // until is the date that the user is suspended until.
      until: {
        type: Date,
        default: null,
      },
      history: [{

        // until is the date that the user is suspended until.
        until: Date,

        // assigned_by stores the user id of the user who assigned this status.
        assigned_by: {type: String, default: null},

        // message stores the email content sent to the user.
        message: {type: String, default: null},

        // created_at stores the date when this status was assigned.
        created_at: {type: Date, default: Date.now}
      }]
    }
  },
```

Reply:
Should be sth like `{status: {banned: {status: status}}` no?